File: java/Tutorials/Media/C/WsqToNImage/WsqToNImage.c
#include <TutorialUtils.h>

#ifdef N_MAC_OSX_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("WsqToNImage");
const NChar description[] = N_T("Demonstrates WSQ to NImage conversion");
const NChar version[] = N_T("9.0.0.0");
const NChar copyright[] = N_T("Copyright (C) 2013-2017 Neurotechnology");

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [srcImage] [dstImage]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\tsrcImage - filename of source WSQ image.\n"));
	printf(N_T("\tdstImage - name of a file to save converted image to.\n"));
	printf(N_T("\n"));
	return 1;
}

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	const NChar * components = N_T("Images.WSQ");
	NBool available = NFalse;
	HNImage hImage = NULL;
	HNImageFormat hImageFormat = NULL;
	HNImageFormat hDstImageFormat = NULL;
	HNImageInfo hImageInfo = NULL;
	NFloat bitRate = 0;

	OnStart(title, description, version, copyright, argc, argv);

	if (argc < 3)
	{
		OnExit();
		return usage();
	}

	// check the license first
	result = NLicenseObtainComponents(N_T("/local"), N_T("5000"), components, &available);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NLicenseObtainComponents() failed, result = %d\n"), result);
		goto FINALLY;
	}

	if (!available)
	{
		printf(N_T("Licenses for %s not available\n"), components);
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	// get WSQ image format
	result = NImageFormatGetWsqEx(&hImageFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NImageFormatGetWsqEx() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// create an NImage from a WSQ image file
	result = NImageCreateFromFileEx(argv[1], hImageFormat, 0, NULL, &hImage);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NImageCreateFromFileEx() failed, result = %d\n"), result);
		goto FINALLY;
	}

	result = NImageGetInfo(hImage, &hImageInfo);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NImageGetInfo() failed, result = %d\n"), result);
		goto FINALLY;
	}

	result = WsqInfoGetBitRate((HWsqInfo)hImageInfo, &bitRate);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("WsqInfoGetBitRate() failed, result = %d\n"), result);
		goto FINALLY;
	}
	printf(N_T("loaded wsq bitrate: %.2f\n"), bitRate);

	// pick a format to save in, e.g. JPEG
	result = NImageFormatGetJpegEx(&hDstImageFormat);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NImageFormatGetJpegEx() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// save image to specified file
	result = NImageSaveToFileEx(hImage, argv[2], hDstImageFormat, NULL, 0);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NImageSaveToFileEx() failed, result = %d\n"), result);
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2;

		result2 = NObjectSet(NULL, &hImage);
		if (NFailed(result2))
			result = PrintErrorMsgWithLastError(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NObjectSet(NULL, &hImageFormat);
		if (NFailed(result2))
			result = PrintErrorMsgWithLastError(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NObjectSet(NULL, &hDstImageFormat);
		if (NFailed(result2))
			result = PrintErrorMsgWithLastError(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NObjectSet(NULL, &hImageInfo);
		if (NFailed(result2))
			result = PrintErrorMsgWithLastError(N_T("NObjectSet() failed, result = %d\n"), result2);
		result2 = NLicenseReleaseComponents(components);
		if (NFailed(result2))
			result = PrintErrorMsgWithLastError(N_T("NLicenseReleaseComponents() failed, result = %d\n"), result2);
	}

	OnExit();

	return result;
}

File: java/Tutorials/Media/Java/wsq-to-nimage/src/main/java/com/neurotec/tutorials/WSQToNImage.java
package com.neurotec.tutorials;

import java.io.IOException;

import com.neurotec.images.NImage;
import com.neurotec.images.NImageFormat;
import com.neurotec.images.WSQInfo;
import com.neurotec.lang.NThrowable;
import com.neurotec.licensing.NLicense;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public final class WSQToNImage {

	private static final String DESCRIPTION = "Demonstrates WSQ to NImage conversion.";
	private static final String NAME = "wsq-to-nimage";
	private static final String VERSION = "9.0.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [srcImage] [dstImage]%n", NAME);
		System.out.println();
		System.out.println("\tsrcImage - filename of source WSQ image.");
		System.out.println("\tdstImage - name of a file to save converted image to.");
		System.out.println();
	}

	public static void main(String[] args) {
		final String components = "Images.WSQ";

		LibraryManager.initLibraryPath();

		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length < 2) {
			usage();
			System.exit(-1);
		}

		NImage image = null;
		try {
			// Obtain license
			if (!NLicense.obtainComponents("/local", 5000, components)) {
				System.err.format("Could not obtain licenses for components: %s%n", components);
				System.exit(-1);
			}

			// Create an NImage from a WSQ image file
			image = NImage.fromFile(args[0], NImageFormat.getWSQ());

			System.out.println("Loaded WSQ bitrate: " + ((WSQInfo) image.getInfo()).getBitRate());

			// Pick a format to save in, e.g. JPEG
			NImageFormat dstFormat = NImageFormat.getJPEG();

			// Save image to specified file
			image.save(args[1], dstFormat);
			System.out.println(dstFormat.getName() + " image was saved to " + args[1]);
		} catch (Throwable th) {
			th.printStackTrace();

			int errorCode = -1;

			if (th instanceof NThrowable) {
				errorCode = ((NThrowable)th).getCode();
			}

			System.exit(errorCode);
		} finally {
			try {
				NLicense.releaseComponents(components);
			} catch (IOException e) {
				e.printStackTrace();
			}
			if (image != null) { image.dispose(); }
		}

	}
}
